tests/0101-fetch-from-follower.cpp

```diff
@@ -0,0 +1,294 @@
+
+  /* arrange for the consumer's client.rack to align with a broker that is not the leader. */
+  for (i=0; i<replica_ids.size(); ++i) {
+    if (replica_ids[i] != leader_id) {
+      client_rack_id = replica_ids[i];
+      break;
+    }
+  }
+  client_rack = tostr() << "RACK" << client_rack_id;
+  Test::Say(tostr() << "client.rack: " << client_rack << "\n");
+
+  Test::conf_init(&conf, NULL, 10);
+  Test::conf_set(conf, "group.id", topic_str);
+  Test::conf_set(conf, "auto.offset.reset", "earliest");
+  Test::conf_set(conf, "enable.auto.commit", "false");
+  Test::conf_set(conf, "statistics.interval.ms", "1000");
+  conf->set("event_cb", &ex_event_cb, errstr);
+  Test::conf_set(conf, "client.rack", client_rack);
+  Test::conf_set(conf, "debug", "all");
+
+  c = RdKafka::KafkaConsumer::create(conf, errstr);
+  test_assert(c, "Failed to create KafkaConsumer: " + errstr);
+  delete conf;
+
+  /* Subscribe */
+  topics.push_back(topic_str);
+  if ((err = c->subscribe(topics)))
+    Test::Fail("subscribe failed: " + RdKafka::err2str(err));
+
+  /* Start consuming */
+  Test::Say("Consuming topic " + topic_str + "\n");
+  cnt = 0;
+  while (cnt < msgcnt) {
+    RdKafka::Message *msg = c->consume(tmout_multip(1000));
+    switch (msg->err())
+      {
+      case RdKafka::ERR__TIMED_OUT:
+        break;
+
+      case RdKafka::ERR_NO_ERROR:
+        cnt++;
+        break;
+
+      default:
+        Test::Fail("Consume error: " + msg->errstr());
+        break;
+      }
+
+    delete msg;
+  }
+
+  /* rely on the test timeout to prevent an infinite loop in
+   * the (unlikely) event that the statistics callback isn't
+   * called. */
+  Test::Say("Capturing rxbytes statistics");
+  TestEvent2Cb::should_capture_stats = true;
+  while (!TestEvent2Cb::has_captured_stats) {
+    RdKafka::Message *msg = c->consume(tmout_multip(500));
+    delete msg;
+  }
+
+  for (i=0; i<replica_ids.size(); ++i)
+    Test::Say(tostr() << _C_YEL << "rxbytes for replica on broker " << replica_ids[i] << ": " << TestEvent2Cb::rxbytes[replica_ids[i]]
+                      << (replica_ids[i] == leader_id ? " (leader)" : "")
+                      << (replica_ids[i] == client_rack_id ? " (preferred replica)" : "")
+                      << "\n");
+
+  for (i=0; i<replica_ids.size(); ++i)
+    if (replica_ids[i] != client_rack_id)
+      test_assert(TestEvent2Cb::rxbytes[replica_ids[i]] < TestEvent2Cb::rxbytes[client_rack_id],
+                  "rxbytes was not highest on broker corresponding to client.rack.");
+
+  test_assert(TestEvent2Cb::rxbytes[client_rack_id] > msgcnt * msgsize,
+              tostr() << "expecting rxbytes of client.rack broker to be at least " << msgcnt * msgsize
+                      << " but it was " << TestEvent2Cb::rxbytes[client_rack_id]);
+
+  Test::Say("Done\n");
+
+  // Manual test 1:
```

Comment:
What's the plan with this?

Reply:
it's not really possible to write tests to cover most of the error scenarios. I did some manual tests that included simple modifications to the code in order to verify some behavior, and i thought it worth noting the procedure here.

Reply:
Manual tests, albeit documented, does not provide any value since they're never run again (in practice) and thus do not catch regression.

Try to see if any of these can be materialized as automatic tests.

Reply:
using these notes I was just able to quickly re-verify behavior following recent changes, so they do provide some value :-). but of course very limited compared to automated regression tests since in practice no one will make use of them unless they're explicitly updating the related code.

it would be very difficult to automate this testing. the java implementation of fff is tested exclusively with unit tests.